sound/soc/intel/boards/sof_rt5682.c

```diff
@@ -61,6 +61,23 @@
 #define SOF_MAX98390_TWEETER_SPEAKER_PRESENT	BIT(25)
 #define SOF_RT1019_SPEAKER_AMP_PRESENT	BIT(26)
 
+/* HDMI capture*/
+#define SOF_SSP_HDMI_CAPTURE_PRESENT            BIT(27)
+#define SOF_NO_OF_HDMI_CAPTURE_SSP_SHIFT                28
+#define SOF_NO_OF_HDMI_CAPTURE_SSP_MASK         (GENMASK(29, 28))
+#define SOF_NO_OF_HDMI_CAPTURE_SSP(quirk)       \
+	(((quirk) << SOF_NO_OF_HDMI_CAPTURE_SSP_SHIFT) & SOF_NO_OF_HDMI_CAPTURE_SSP_MASK)
+
+#define SOF_HDMI_CAPTURE_1_SSP_SHIFT            30
+#define SOF_HDMI_CAPTURE_1_SSP_MASK             (GENMASK(32, 30))
```

Comment:
we need to define sof_rt5682_quirk as u64 now... see line 83

Reply:
our observation from test is unsigned long sof_rt5682_quirk is supported 64 bit when its compiled with X64 flag enabled.
the issue is during macro assignment and shift , SOF_HDMI_CAPTURE_1_SSP(quirk)=> quirk is treated as 32 bit and throws complier warning. Hence we need to do temporary type casting to make it as 64 bit variable.

Reply:
![overflow_warning](https://github.com/thesofproject/linux/assets/85683900/32cbc176-d26b-4a36-b7df-12b61c47b0a1)

Reply:
no need for typecast, use u64 for the quirk.

Reply:
I suggested using u64, that gives no ambiguity on the type.

Reply:
Below are the changes we tried but same error. please confirm, Is this a way you have suggested,
<img width="629" alt="Modified_code" src="https://github.com/thesofproject/linux/assets/85683900/a9f0e187-4b34-4eac-a184-7130f71cd7c3">

and same warning,
<img width="682" alt="overflow_warn" src="https://github.com/thesofproject/linux/assets/85683900/e228d0af-2b76-4817-bfb0-793017237d0c">

Reply:
Even this macro typecasting helps only for X64 bit build system. Still SOF CI i386 build fails since it can't handle more than 32 bits. we may need to revisit the sof_rt5682_quirk implementation for above 32 bit handling and compatible to i386 32 bit system builds .

sound/soc/intel/boards/sof_rt5682.c: In function ‘sof_card_dai_links_create’:
./include/linux/bits.h:35:11: error: right shift count is negative [-Werror=shift-count-negative]
   35 |   (~UL(0) >> (BITS_PER_LONG - 1 - (h))))
      |           ^~
./include/linux/bits.h:37:31: note: in expansion of macro ‘__GENMASK’
   37 |  (GENMASK_INPUT_CHECK(h, l) + __GENMASK(h, l))
      |                               ^~~~~~~~~
sound/soc/intel/boards/sof_rt5682.c:72:50: note: in expansion of macro ‘GENMASK’
   72 | #define SOF_HDMI_CAPTURE_1_SSP_MASK             (GENMASK(32, 30))
      |                                                  ^~~~~~~
sound/soc/intel/boards/sof_rt5682.c:884:45: note: in expansion of macro ‘SOF_HDMI_CAPTURE_1_SSP_MASK’
  884 |    int port = (i == 1 ? (sof_rt5682_quirk & SOF_HDMI_CAPTURE_1_SSP_MASK) >>
      |                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~
./include/linux/bits.h:34:22: error: left shift count >= width of type [-Werror=shift-count-overflow]
   34 |  (((~UL(0)) - (UL(1) << (l)) + 1) & \
      |                      ^~
./include/linux/bits.h:37:31: note: in expansion of macro ‘__GENMASK’
   37 |  (GENMASK_INPUT_CHECK(h, l) + __GENMASK(h, l))
      |                               ^~~~~~~~~
sound/soc/intel/boards/sof_rt5682.c:77:50: note: in expansion of macro ‘GENMASK’
   77 | #define SOF_HDMI_CAPTURE_2_SSP_MASK             (GENMASK(35, 33))
      |                                                  ^~~~~~~
sound/soc/intel/boards/sof_rt5682.c:886:27: note: in expansion of macro ‘SOF_HDMI_CAPTURE_2_SSP_MASK’
  886 |       (sof_rt5682_quirk & SOF_HDMI_CAPTURE_2_SSP_MASK) >>

Reply:
> no need for typecast, use u64 for the quirk.

@plbossart the quirk is a number when `SOF_HDMI_CAPTURE_2_SSP` is used. For example,
```
.driver_data = (kernel_ulong_t)(SOF_RT5682_MCLK_EN |
                                        SOF_RT5682_SSP_CODEC(1) |
                                        SOF_RT5682_NUM_HDMIDEV(3) |
                                        SOF_NO_OF_HDMI_CAPTURE_SSP(2) |
                                        SOF_HDMI_CAPTURE_1_SSP(0) |
                                        SOF_HDMI_CAPTURE_2_SSP(2) |
                                        SOF_SSP_HDMI_CAPTURE_PRESENT),
```
And the number is considered to be int. So we need to cast the number to u64, like `SOF_HDMI_CAPTURE_2_SSP((u64)2)`.
IMHO, it makes sense to add the typecasting in the definition.

Reply:
Hi @plbossart @bardliao,

Please provide your suggestion that whether we can ignore SOF CI i386 build failure. Since it can't support above 32 bit and in future also if we want add more quirk bits then its always blocker.

Reply:
> Hi @plbossart @bardliao,
> 
> Please provide your suggestion that whether we can ignore SOF CI i386 build failure. Since it can't support above 32 bit and in future also if we want add more quirk bits then its always blocker.

@CBala21 I think we can't ignore SOF CI i386 build failure. Otherwise, it will be always failed after the PR is merged. I would suggest that you use a 4 bits mask to present HDMI IN. 
#define SOF_SSP_HDMI_CAPTURE_PRESENT_MASK   (GENMASK(30, 27))
And you can get all information you need from the quirk.
unsigned int hdmi_in = (sof_rt5682_quirk & SOF_SSP_HDMI_CAPTURE_PRESENT_MASK) >>
				SOF_NO_OF_HDMI_CAPTURE_SSP_SHIFT;
hweight32(hdmi_in) = how many hdmi in do you have.
 for_each_set_bit(bit, &hdmi_in, 32) {} can know which SSP ports are used by hdmi in.
For example, the hdmi_in should be 0x5, i.e. bit 0 and 2 is set when SSP 0 and 2 are used by HDMI in.
And  hdmi_in should be 0x0 if there is no HDMI in.
What do you think?

Reply:
Hi @bardliao,

As per your suggestion changed the code and uploaded. please check.

Reply:
@CBala21 Thanks, it looks good. Can you split the commit into 3?
1. Add HDMI_IN SSP support
2. Add adl_rt5682_c1_h02 id
3. Add adl_rt5682_rt5682s_hp in snd_soc_acpi_intel_adl_machines[]

Reply:
Hi [bardliao],

All these three changes are connected without this HDMI -in capture use case wont work. Still you want to split it out?